File: dart/design_patterns/lib/structural/composite.dart
// Composite Pattern

// Component
abstract class FileComponent {
  final String name;

  FileComponent(this.name);

  int getSize();
  void printInfo();
}

// Leaf
class File extends FileComponent {
  final int _sizeInBytes;

  File(super.name, this._sizeInBytes);

  @override
  int getSize() => _sizeInBytes;

  @override
  void printInfo() {
    print('--- file $name size=${getSize()} bytes');
  }
}

// Composite
class Directory extends FileComponent {
  final List<FileComponent> _components = [];

  Directory(super.name);

  void addComponent(FileComponent component) {
    _components.add(component);
  }

  @override
  int getSize() => _components.fold(0, (sum, c) => sum + c.getSize());

  @override
  void printInfo() {
    print('-- dir $name size=${getSize()} bytes');
    for (final component in _components) {
      component.printInfo();
    }
  }
}

// Usage
void compositeExample() {
  final file1 = File('document.txt', 1024);
  final file2 = File('image.png', 2048);
  final file3 = File('notes.md', 512);

  final subDir = Directory('subfolder')
    ..addComponent(file2)
    ..addComponent(file3);

  final rootDir = Directory('root')
    ..addComponent(file1)
    ..addComponent(subDir);

  rootDir.printInfo();
}

File: swift/DesignPatterns/Structural/Composite.swift
//
//  Composite.swift
//  DesignPatterns
//
//  Created by Soe Min Thein on 23/03/2026.
//

import Foundation

// MARK: - Component

protocol FileComponent {
    var name: String { get }
    func getSize() -> Int
    func printInfo()
}

// MARK: - Leaf

class File: FileComponent {
    let name: String
    private let sizeInBytes: Int

    init(name: String, sizeInBytes: Int) {
        self.name = name
        self.sizeInBytes = sizeInBytes
    }

    func getSize() -> Int {
        return sizeInBytes
    }

    func printInfo() {
        print("--- file \(name) size=\(getSize()) bytes")
    }
}

// MARK: - Composite

class Directory: FileComponent {
    let name: String
    private var components: [FileComponent] = []

    init(name: String) {
        self.name = name
    }

    func addComponent(_ component: FileComponent) {
        components.append(component)
    }

    func getSize() -> Int {
        return components.reduce(0) { $0 + $1.getSize() }
    }

    func printInfo() {
        print("-- dir \(name) size=\(getSize()) bytes")
        for component in components {
            component.printInfo()
        }
    }
}

// MARK: - Usage

func CompositeExample() {
    
    let file1 = File(name: "document.txt", sizeInBytes: 1024)
    let file2 = File(name: "image.png", sizeInBytes: 2048)
    let file3 = File(name: "notes.md", sizeInBytes: 512)
    
    let subDir = Directory(name: "subfolder")
    subDir.addComponent(file2)
    subDir.addComponent(file3)
    
    let rootDir = Directory(name: "root")
    rootDir.addComponent(file1)
    rootDir.addComponent(subDir)
    
    rootDir.printInfo()
}
